File: backend/services/ai-service/src/modules/ai/prompts/askQuestion.ts
import type { AskQuestionInput, LearnerContext } from '../ai.types';

function buildLearnerProfileSection(ctx: LearnerContext): string {
  const parts: string[] = [];

  const level = ctx.familiarityLevel ?? 'beginner';
  parts.push(`Level: ${level}`);

  if (ctx.learningGoal) parts.push(`Goal: ${ctx.learningGoal}`);
  if (ctx.preferredLearningStyle) parts.push(`Preferred style: ${ctx.preferredLearningStyle}`);
  if (ctx.priorSkills) parts.push(`Prior skills: ${ctx.priorSkills}`);
  if (ctx.aboutSelf) parts.push(`About them: ${ctx.aboutSelf}`);

  if (ctx.currentNodeAttempts > 0) {
    const score = ctx.currentNodeBestScore !== null ? ` (best: ${ctx.currentNodeBestScore}%)` : '';
    parts.push(`Attempts on this node: ${ctx.currentNodeAttempts}${score}`);
  }

  if (ctx.totalNodes > 0) {
    parts.push(`Progress: ${ctx.nodesCompleted}/${ctx.totalNodes} nodes completed`);
  }

  return parts.join('\n');
}

export function buildAskPrompt(input: AskQuestionInput): string {
  const { nodeTitle, description, learningOutcomes, question, explanation, learnerContext } = input;

  const contextParts: string[] = [];

  if (description) {
    contextParts.push(`Description: ${description}`);
  }

  if (learningOutcomes && learningOutcomes.length > 0) {
    contextParts.push(`Learning outcomes:\n${learningOutcomes.map((o) => `  - ${o}`).join('\n')}`);
  }

  if (explanation) {
    contextParts.push(`AI-generated summary: ${explanation.summary}`);
    if (explanation.keyPoints.length > 0) {
      contextParts.push(
        `Key concepts covered:\n${explanation.keyPoints.map((p) => `  - ${p}`).join('\n')}`,
      );
    }
    if (explanation.commonMistakes && explanation.commonMistakes.length > 0) {
      contextParts.push(
        `Common mistakes to avoid:\n${explanation.commonMistakes.map((m) => `  - ${m}`).join('\n')}`,
      );
    }
  }

  const learnerBlock = learnerContext
    ? `\nLearner profile:\n${buildLearnerProfileSection(learnerContext)}\n`
    : '';

  const depthGuidance = learnerContext?.familiarityLevel === 'advanced'
    ? 'Give a precise, technical answer. Skip basics they already know.'
    : learnerContext?.familiarityLevel === 'intermediate'
      ? 'Balance clarity with technical depth.'
      : 'Use simple language and concrete analogies. Be encouraging.';

  return `You are a knowledgeable and encouraging AI instructor helping a learner study "${nodeTitle}".

${contextParts.length > 0 ? `Node context:\n${contextParts.join('\n\n')}\n\n` : ''}${learnerBlock}
Student's question: ${question}

${depthGuidance}
Answer clearly and helpfully in 2-5 sentences. Tailor depth to the question and learner profile.
Return ONLY this JSON with no extra text:
{"answer":"<your answer here>"}`;
}

/**
 * Streaming variant — identical context but instructs the model to respond
 * in plain readable text so tokens can be streamed word-by-word.
 */
export function buildStreamAskPrompt(input: AskQuestionInput): string {
  const { nodeTitle, description, learningOutcomes, question, explanation, learnerContext } = input;

  const contextParts: string[] = [];

  if (description) {
    contextParts.push(`Description: ${description}`);
  }

  if (learningOutcomes && learningOutcomes.length > 0) {
    contextParts.push(`Learning outcomes:\n${learningOutcomes.map((o) => `  - ${o}`).join('\n')}`);
  }

  if (explanation) {
    contextParts.push(`AI-generated summary: ${explanation.summary}`);
    if (explanation.keyPoints.length > 0) {
      contextParts.push(
        `Key concepts covered:\n${explanation.keyPoints.map((p) => `  - ${p}`).join('\n')}`,
      );
    }
    if (explanation.commonMistakes && explanation.commonMistakes.length > 0) {
      contextParts.push(
        `Common mistakes to avoid:\n${explanation.commonMistakes.map((m) => `  - ${m}`).join('\n')}`,
      );
    }
  }

  const learnerBlock = learnerContext
    ? `\nLearner profile:\n${buildLearnerProfileSection(learnerContext)}\n`
    : '';

  const depthGuidance = learnerContext?.familiarityLevel === 'advanced'
    ? 'Give a precise, technical answer. Skip basics they already know.'
    : learnerContext?.familiarityLevel === 'intermediate'
      ? 'Balance clarity with technical depth.'
      : 'Use simple language and concrete analogies. Be encouraging.';

  return `You are a knowledgeable and encouraging AI instructor helping a learner study "${nodeTitle}".

${contextParts.length > 0 ? `Node context:\n${contextParts.join('\n\n')}\n\n` : ''}${learnerBlock}
Student's question: ${question}

${depthGuidance}
Answer clearly and helpfully in 2-5 sentences. Tailor depth to the question and learner profile.
Write your answer as plain readable text. Do not use JSON, code blocks, or any special formatting.`;
}

File: backend/services/ai-service/app/prompts/ask_question.py
from app.schemas import AskQuestionInput, LearnerContext


def _learner_profile_section(ctx: LearnerContext) -> str:
    parts: list[str] = []
    level = ctx.familiarity_level or "beginner"
    parts.append(f"Level: {level}")
    if ctx.learning_goal:
        parts.append(f"Goal: {ctx.learning_goal}")
    if ctx.preferred_learning_style:
        parts.append(f"Preferred style: {ctx.preferred_learning_style}")
    if ctx.prior_skills:
        parts.append(f"Prior skills: {ctx.prior_skills}")
    if ctx.about_self:
        parts.append(f"About them: {ctx.about_self}")
    if ctx.current_node_attempts > 0:
        score = f" (best: {ctx.current_node_best_score}%)" if ctx.current_node_best_score is not None else ""
        parts.append(f"Attempts on this node: {ctx.current_node_attempts}{score}")
    if ctx.total_nodes > 0:
        parts.append(f"Progress: {ctx.nodes_completed}/{ctx.total_nodes} nodes completed")
    return "\n".join(parts)


def _context_block(input_data: AskQuestionInput) -> str:
    parts: list[str] = []
    if input_data.description:
        parts.append(f"Description: {input_data.description}")
    if input_data.learning_outcomes:
        items = "\n".join(f"  - {o}" for o in input_data.learning_outcomes)
        parts.append(f"Learning outcomes:\n{items}")
    if input_data.explanation:
        exp = input_data.explanation
        parts.append(f"AI-generated summary: {exp.summary}")
        if exp.key_points:
            items = "\n".join(f"  - {p}" for p in exp.key_points)
            parts.append(f"Key concepts covered:\n{items}")
        if exp.common_mistakes:
            items = "\n".join(f"  - {m}" for m in exp.common_mistakes)
            parts.append(f"Common mistakes to avoid:\n{items}")
    return "\n\n".join(parts)


def _depth_guidance(ctx: LearnerContext | None) -> str:
    level = ctx.familiarity_level if ctx else "beginner"
    if level == "advanced":
        return "Give a precise, technical answer. Skip basics they already know."
    elif level == "intermediate":
        return "Balance clarity with technical depth."
    return "Use simple language and concrete analogies. Be encouraging."


def build_ask_prompt(input_data: AskQuestionInput) -> str:
    ctx_block = _context_block(input_data)
    learner_block = f"\nLearner profile:\n{_learner_profile_section(input_data.learner_context)}\n" if input_data.learner_context else ""
    depth = _depth_guidance(input_data.learner_context)
    ctx_section = f"Node context:\n{ctx_block}\n\n" if ctx_block else ""

    return f"""You are a knowledgeable and encouraging AI instructor helping a learner study "{input_data.node_title}".

{ctx_section}{learner_block}
Student's question: {input_data.question}

{depth}
Answer clearly and helpfully in 2-5 sentences. Tailor depth to the question and learner profile.
Return ONLY this JSON with no extra text:
{{"answer":"<your answer here>"}}"""


def build_stream_ask_prompt(input_data: AskQuestionInput) -> str:
    """Streaming variant — plain text output so tokens stream naturally."""
    ctx_block = _context_block(input_data)
    learner_block = f"\nLearner profile:\n{_learner_profile_section(input_data.learner_context)}\n" if input_data.learner_context else ""
    depth = _depth_guidance(input_data.learner_context)
    ctx_section = f"Node context:\n{ctx_block}\n\n" if ctx_block else ""

    return f"""You are a knowledgeable and encouraging AI instructor helping a learner study "{input_data.node_title}".

{ctx_section}{learner_block}
Student's question: {input_data.question}

{depth}
Answer clearly and helpfully in 2-5 sentences. Tailor depth to the question and learner profile.
Write your answer as plain readable text. Do not use JSON, code blocks, or any special formatting."""
